### DISCUSSION TITLE
How is Kaboom.js an ECS?

### DISCUSSION THREAD

**Discussion by ghost** [replit/kaboom#619] [General]
I am curious on how Kaboom.js an Entity Component System (ECS)? When seeing the examples, I can see that the library has two parts of an ECS. These parts are the entity (created via `add()`), and the component (ex. `body()`, `health()`,  `sprite()`, etc.). The only part I cannot see based on the API is the system. How does the library know what to render and how to render it? How does the library know what entities need to be checked for collision detection? Are new actions/abilities added with new components rather than with new systems?

**Comment by slmjkdbtl** [comment]
Kaboom takes ideas from ECS but isn't an ECS. The logic and data is still coupled in components and there's no concept of a system, prioritizing ease of use.

Kaboom has lower level drawing APIs like `drawSprite()`, `drawRect()` etc similar to the p5.js / love2d / canvas2d drawing functions, components like `sprite()` will call these in the `draw()` event to do the rendering. You can also use `drawSprite()` etc yourself in `onDraw()` if you want to do something custom.
